libs/content/workspace/src/tab/svg_editor/renderer.rs

```diff
@@ -0,0 +1,268 @@
+                } else {
+                    true
+                }
+            }));
+        }
+
+        buffer
+            .elements
+            .iter_mut()
+            .for_each(|(_, element)| match element {
+                parser::Element::Path(p) => p.diff_state = DiffState::default(),
+                parser::Element::Image(i) => i.diff_state = DiffState::default(),
+                parser::Element::Text(_) => todo!(),
+            })
+    }
+}
+
+fn load_image_textures(buffer: &mut Buffer, ui: &mut egui::Ui) {
+    for (id, el) in buffer.elements.iter_mut() {
+        if let parser::Element::Image(img) = el {
+            match &img.data {
+                ImageKind::JPEG(bytes) | ImageKind::PNG(bytes) => {
+                    let image = image::load_from_memory(bytes).unwrap();
+
+                    let egui_image = egui::ColorImage::from_rgba_unmultiplied(
+                        [image.width() as usize, image.height() as usize],
+                        &image.to_rgba8(),
+                    );
+
+                    if img.texture.is_none() {
+                        img.texture = Some(ui.ctx().load_texture(
+                            format!("canvas_img_{}", id),
+                            egui_image,
+                            egui::TextureOptions::LINEAR,
+                        ));
+                    }
+                }
+                ImageKind::GIF(_) => todo!(),
+                ImageKind::SVG(_) => todo!(),
+            }
+        }
+    }
+}
+
+// todo: maybe impl this on element struct
+fn tesselate_element(
+    el: &mut parser::Element, id: &String, dark_mode: bool, master_transform: Transform,
+) -> Option<(String, RenderOp)> {
+    let mut mesh: VertexBuffers<_, u32> = VertexBuffers::new();
+    let mut stroke_tess = StrokeTessellator::new();
+
+    match el {
+        parser::Element::Path(p) => {
+            if let Some(stroke) = p.stroke {
+                if p.data.is_empty() {
+                    return None;
+                }
+                let stroke_color = if dark_mode { stroke.color.1 } else { stroke.color.0 }
+                    .linear_multiply(p.opacity);
+
+                let mut builder = lyon::path::BuilderWithAttributes::new(1);
+
+                let mut first = None;
+                let mut i = 0;
+
+                while let Some(seg) = p.data.get_segment(i) {
+                    let pressure = if let Some(ref pressure) = p.pressure {
+                        let pressure_at_segment = pressure.get(i);
+                        if pressure_at_segment.is_some() {
+                            pressure_at_segment
+                        } else {
+                            pressure.get(i - 1)
+                        }
+                    } else {
+                        None
+                    };
+
+                    let normalized_pressure =
+                        if let Some(prsr) = pressure { *prsr + 0.5 } else { 1.0 };
```

Comment:
How'd you come up with `+ 0.5`? When I worked on the C# Windows client and tried to make drawings show up the same there as it did on Apple I just measured the pixel width of strokes of various pressure ended up with a `+ 0.16`, a `* 1.55`, and a clamp between 0.16 and 1.0

Reply:
https://github.com/lockbook/lockbook/commit/2991203d864fbb76972b96864817efbd0768cc06#diff-c2e5e440a219125a999af52a9e24f374447067b9411e86b3f01ce74060cf9e25 (clients/windows/core/Draw.cs if the link doesn't work)

Reply:
worth calling out that ideally pressure will be a device specific knob which will result in a thickness stored in some svg complaint manner. It's a slightly different process as we'll likely be trying to match the pen *behavior* of the device, rather than lockbook's appearance from device to device

Reply:
@tvanderstad  egui expects the force value to be normalized [between 0 and 1](https://docs.rs/egui/latest/egui/enum.Event.html#variant.Touch).
in this case i'm shifting the range form [0,1] to [0.5,1.5] because when you press very lightly it'll be jarring to have the thickness multiplied by a value very close to 0 and cause it to almost disappear from the screen.